### PR TITLE
httpserver: Makeover

### PR DESCRIPTION
**Pull request by efritz**
Reorganize the httpserver package so that we can supply listeners, pre-constructed handlers and server options instead of just a port number.

This adapts the httpserver as something that can run the frontend servers, and gives us a good set of extension points to ensure we start servers in different services in a similar way (a path forward to install middleware on all service request handlers, etc).

### CODE REVIEW

File: enterprise/cmd/executor-queue/main.go [hunk 1]
```diff
@@ -68,10 +68,15 @@ func main() {
 		}))
 	}
 
+	server, err := apiserver.NewServer(serviceConfig.ServerOptions(queueOptions))
+	if err != nil {
+		log.Fatalf("failed to create listener: %s", err)
```
Comment (eseliger) [inline enterprise/cmd/executor-queue/main.go:73]: shouldn't this fail? Or does log.Fatalf panic?

File: enterprise/cmd/executor/main.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/workerutil"
 )
 
-const port = 3192
+const port = ":3192"
```
Comment (eseliger) [inline enterprise/cmd/executor/main.go:22]: :( to me, the only real ports are ints 😬 

(feel free to ignore)

File: enterprise/cmd/precise-code-intel-worker/main.go [hunk 3]
```diff
@@ -83,8 +85,14 @@ func main() {
...
 
-	go debugserver.Start()
-	goroutine.MonitorBackgroundRoutines(context.Background(), server, uploadResetter, commitUpdater, worker)
+	goroutine.MonitorBackgroundRoutines(
+		context.Background(),
+		server,
+		uploadResetter,
+		commitUpdater,
+		worker,
+		goroutine.NoopStop(debugserver.NewServerRoutine()),
```
Comment (eseliger) [inline enterprise/cmd/precise-code-intel-worker/main.go:94]: A server needs no "tear-down" method? 
I think IIRC, ports are sometimes only slowly released when not properly returned, at least when using the C standard library, but maybe the go runtime is doing it better.

**Review by eseliger — APPROVED** [review]

File: enterprise/cmd/executor-queue/main.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/executor-queue/main.go:73]: "Fatal" implies exit the program.

File: enterprise/cmd/precise-code-intel-worker/main.go (hunk 3 above)
Comment (unknwon) [reply enterprise/cmd/precise-code-intel-worker/main.go:94]: From my experience, ports are released "fast enough" after program exited, never proved though.

File: enterprise/cmd/executor/main.go (hunk 2 above)
Comment (unknwon) [reply enterprise/cmd/executor/main.go:22]: Technically... I think these consts should be renamed to `Addr`.

File: internal/httpserver/server.go [hunk 4]
```diff
@@ -5,49 +5,51 @@ import (
...
-		server: &http.Server{
-			Addr:    net.JoinHostPort(host, strconv.FormatInt(int64(port), 10)),
-			Handler: ot.Middleware(router),
-		},
+		listener: listener,
+		server:   httpServer,
 	}
 }
 
+// New returns a BackgroundRoutine that serves the given handler on the given port.
```
Comment (unknwon) [inline internal/httpserver/server.go:41]:
```suggestion
// New returns a BackgroundRoutine that serves the given handler on the given address.
```

File: enterprise/cmd/precise-code-intel-worker/main.go [hunk 5]
```diff
@@ -62,7 +61,10 @@ func main() {
 	MustRegisterQueueMonitor(observationContext.Registerer, store)
 	workerMetrics := metrics.NewWorkerMetrics(observationContext)
 	resetterMetrics := resetter.NewResetterMetrics(prometheus.DefaultRegisterer)
-	server := httpserver.New(Port, func(router *mux.Router) {})
+	server, err := httpserver.NewFromAddr(port, httpserver.NewHandler(nil), httpserver.Options{})
+	if err != nil {
+		log.Fatalf("failed to create listener: %s", err)
```
Comment (unknwon) [inline enterprise/cmd/precise-code-intel-worker/main.go:66]: nit:
```suggestion
		log.Fatalf("Failed to create listener: %s", err)
```

File: enterprise/cmd/precise-code-intel-indexer/main.go [hunk 6]
```diff
@@ -67,7 +67,10 @@ func main() {
 		UnreportedIndexMaxAge: cleanupInterval * time.Duration(maximumMissedHeartbeats),
 		DeathThreshold:        cleanupInterval * time.Duration(maximumMissedHeartbeats),
 	})
-	server := server.New(indexManager)
+	server, err := server.New(indexManager)
+	if err != nil {
+		log.Fatalf("failed to create listener: %s", err)
```
Comment (unknwon) [inline enterprise/cmd/precise-code-intel-indexer/main.go:72]: nit:
```suggestion
		log.Fatalf("Failed to create listener: %s", err)
```

File: enterprise/cmd/precise-code-intel-indexer/internal/server/server.go [hunk 7]
```diff
@@ -6,16 +6,16 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/httpserver"
 )
 
-const Port = 3189
+const port = ":3189"
```
Comment (unknwon) [inline enterprise/cmd/precise-code-intel-indexer/internal/server/server.go:9]:
```suggestion
const addr = ":3189"
```

File: enterprise/cmd/precise-code-intel-bundle-manager/main.go [hunk 8]
```diff
@@ -78,12 +78,16 @@ func main() {
 	store := store.NewObserved(mustInitializeStore(), observationContext)
 	metrics.MustRegisterDiskMonitor(bundleDir)
 
-	server := server.New(bundleDir, storeCache, codeIntelDB, observationContext)
+	server, err := server.New(bundleDir, storeCache, codeIntelDB, observationContext)
+	if err != nil {
+		log.Fatalf("failed to create listener: %s", err)
```
Comment (unknwon) [inline enterprise/cmd/precise-code-intel-bundle-manager/main.go:83]: nit:
```suggestion
		log.Fatalf("Failed to create listener: %s", err)
```

File: enterprise/cmd/precise-code-intel-bundle-manager/internal/server/server.go [hunk 9]
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/observation"
 )
 
-const Port = 3187
+const port = ":3187"
```
Comment (unknwon) [inline enterprise/cmd/precise-code-intel-bundle-manager/internal/server/server.go:12]:
```suggestion
const addr = ":3187"
```

File: enterprise/cmd/executor/main.go [hunk 10]
```diff
@@ -40,7 +40,12 @@ func main() {
 		apiworker.NewWorker(config.APIWorkerOptions(nil)),
 	}
 	if !config.DisableHealthServer {
-		routines = append(routines, httpserver.New(port, nil))
+		server, err := httpserver.NewFromAddr(port, httpserver.NewHandler(nil), httpserver.Options{})
+		if err != nil {
+			log.Fatalf("failed to create listener: %s", err)
```
Comment (unknwon) [inline enterprise/cmd/executor/main.go:45]: nit:
```suggestion
			log.Fatalf("Failed to create listener: %s", err)
```

**Review by unknwon — APPROVED** [review]
Approve to unblock.

File: enterprise/cmd/precise-code-intel-worker/main.go (hunk 3 above)
Comment (efritz) [reply enterprise/cmd/precise-code-intel-worker/main.go:94]: I'm actually handling this as part of another effort. Stay tuned!

**Pull request merged**